### PR TITLE
EVG-14112: Introduce post_error_fails_task

### PR DESCRIPTION
**Pull request by hadjri**
[EVG-14112](https://jira.mongodb.org/browse/EVG-14112)

### Description 
Currently failures in post task commands are ignored by default.  A change has been introduced to allow users to enforce in config that failures during post tasks cause the task to fail (similar to [pre_error_fails_task](https://github.com/evergreen-ci/evergreen/wiki/Project-Configuration-Files#pre-post-and-timeout))

### Testing 
Added unit tests to ensure that these new configurations let post tasks to return an error if they are set.

### CODE REVIEW

File: model/project.go
Comment (ablack12) [inline model/project.go:495]: TeardownGroup is actually something else and won't work for this, so this is misleading. It's actually TeardownTask (i.e. post tasks) that we're allowing to fail (not the best naming scheme)

**Review by ablack12 — CHANGES_REQUESTED** [review]
You should be able to test this in staging by adding a post task with a shell.exec command "exit 1" (which will always fail), and then verify that post_error_fails_task does cause the task to fail (and that it doesn't fail without that)

File: agent/agent.go
Comment (hadjri) [reply agent/agent.go:641]: shouldSetupFail is part of the runCommandsOptions struct.  Currently it is set as flagged to true if pre_error_fails_task is true because this will set SetupGroupFailTask and this property is referring to that. We can introduce a new property in the struct called something like shouldTeardownFail, or we could rename this property to something more generic that can apply to pre and post task commands, like shouldTriggerFailure.

**Comment by hadjri** [comment]
> You should be able to test this in staging by adding a post task with a shell.exec command "exit 1" (which will always fail), and then verify that post_error_fails_task does cause the task to fail (and that it doesn't fail without that)

Tested in staging and observed that once a post task fails, the entire task will fail if post_error_fails_task is set: https://spruce-staging.corp.mongodb.com/task/evg_ubuntu1604_test_model_event_patch_db3f306dac84b094a7a55942a0c6f03995a916a3_6111591697b1d349228bf2dc_21_08_09_16_34_46/logs?execution=0&sortBy=STATUS&sortDir=ASC

Compared to currently in prod all post tasks will execute and do not have an impact on whether the task fails or not: https://evergreen.mongodb.com/task/evergreen_race_detector_test_model_event_8b69f0ddb6666271114214e0a9b57620a2e27e61_21_08_09_16_18_52

File: agent/agent.go
Comment (hadjri) [reply agent/agent.go:646]: Can you specify where the double logging occurs? It looks like runCommands only returns an error and doesn't log it out, and this log message has been in the codebase for some time.

File: config.go
Comment (ablack12) [inline config.go:91]: Is this from the merge?

File: agent/agent.go
Comment (ablack12) [inline agent/agent.go:643]: (moving conversation from a previous thread here, since it's attached to outdated code)
I think we should rename this field then. It isn't going to be clear to us in the future why teardown is affecting setup (or so it reads here).

File: agent/agent.go
Comment (ablack12) [reply agent/agent.go:646]: Oh I think I said the wrong thing but there is still double logging, because we're logging here, and we're logging in FinishTask if runPostTaskCommands returns an error (I think we can just remove that log)

File: agent/agent.go
Comment (ablack12) [inline agent/agent.go:553]: I think the issue with this is that most users won't be using post_error_fails_task, but now they suddenly have to wait for post task to know if the task has been successful or not. Maybe the more consistent user experience would be to log this before post task still, and then log "Post task completed -- FAILURE" if we're allowing that (this may be mildly confusing for that use case but it'll be more rare anyway)

**Review by ablack12 — CHANGES_REQUESTED** [review]
Were you able to test this in staging? I would like to verify that it works end to end before merging.

File: config.go
Comment (hadjri) [reply config.go:91]: Yes, looks like something didn't go totally right when I was resolving merge conflicts in this file.  Fixing.

**Comment by hadjri** [comment]
> Were you able to test this in staging? I would like to verify that it works end to end before merging.

Tested in staging with a post command (shell.exec) just running exit 1 in script.  The task is marked as failed even though all tests passed due to this post task failing alongside post_err_fails_task being enabled.  Will also update documentation when change is merged. https://spruce-staging.corp.mongodb.com/task/evg_ubuntu1604_test_model_event_patch_c6ffa29e590ee1ce9137bbd6827cc2767feff144_611289ad9dbe3250b56347ec_21_08_10_14_14_50/logs?execution=0

File: model/project.go
Comment (ablack12) [inline model/project.go:495]: nit: can you keep the setupgroup variables together and move this before or after (or next to teardown)

**Review by ablack12 — CHANGES_REQUESTED** [review]
Sorry for the long wait on the review! Feel free to ping me if it's been more than 24 hours.

This is just about there, we'll just need to add post error to this merge function https://github.com/evergreen-ci/evergreen/blob/cd549ca996e22bf1b812023ed403c19bc84be478/model/project_parser.go#L1207 (there's a ticket planned to create a linter so that we catch when things haven't been added to this merge function in the future)

**Review by ablack12 — APPROVED** [review]

**Comment by hadjri** [comment]
Updated documentation: https://github.com/evergreen-ci/evergreen/wiki/Project-Configuration-Files#pre-post-and-timeout

**Pull request merged**
